Add lint.sh with code to check documentation for URL consistency

joelfranusic-okta opened the pull request:
- Install ruby version found in .ruby-version
- Use rvm as a function instead of a command
- Install bundler again, since we'll have a new version of ruby

joelfranusic-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    echo "----- ${1} -----"
+    }
+
+function setup {
```
This should be `setup()`

joelfranusic-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    exit $BUILD_FAILURE;
+fi
+
+if url_consistency_check;
```
The logic of `url_consistency_check` should be inverted so that this line reads `if ! url_consistency_check` as above.

jmaldonado-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+# Copyright 2016 Okta, Inc.
+
+# Quit if any of the commands in this script fail
+set -e
```
In general, this is not a bad practice but we prefer adding error handlers since this script will be wrapped up by the bacon flow. Thus, it is encouraged by us to disable flag `-e` to avoid killing the script before sending the correct error code and logs to bacon. (The failure code is still raised by the error handler)

rlucero-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    # Note: The file `.ruby-version` has the current ruby version and is used by rbenv
+    # https://rvm.io/workflow/projects#project-file-ruby-version
+    # For example, this file might contain this line: "ruby-2.0.0-p643"
+    rvm install `cat .ruby-version`
```
Double check with Eng-Prod, this shouldn't be required for CI machines.
@tsecor-okta @ajinkyasuryawanshi-okta @kelvinzhu-okta

jmaldonado-okta replied:
Are you guys planning on using Albatross for development? It would be nice if you do so because that would also mean you would be aligned with CI when building locally then `setup()` should not be required for ruby since all docker containers already include a version.

rlucero-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    local status=$?
+    interject 'Done building Jekyll website'
+    return $status
+    }
```
nit. Inconsistent bracket spacing.

rlucero-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+       return 1;
+    fi
+
+    url_consistency_check_file=`mktemp`
```
nit. Can this be a named file that can be pulled up and posted during CI runs?

joelfranusic-okta replied:
I just remembered, the reason why I use `tee` below is so that the contents of this file show up in the logs.

rchild-okta replied:
Docs here:
https://oktawiki.atlassian.net/wiki/display/eng/Using+Bacon+to+build+your+artifact#UsingBacontobuildyourartifact-SettingUpAdditionalLogFiles

An example in the angular repo is here:
https://github.com/okta/samples-js-angular-1/blob/master/scripts/logs

jmaldonado-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    #     0     One or more lines were selected.
+    #     1     No lines were selected.
+    #     >1    An error occurred.
+    grep . $url_consistency_check_file > /dev/null
```
I am confused by what is the PASS criteria. I have a few comments but they depend on what the criteria is since is not clear to me what we are testing.

For example, in the current implementation there could be entries in file `url_consistency_check_file ` that will not necessarily match the regex in the `sed` above or even the file could contain only non-valid entries. Is this ok or are you just loooking for a non-empty file at the end? You probably want to print only matches via `-n` and raise if the regex does not match via `q`

joelfranusic-okta replied:
As far as the `find | xargs grep | grep | sed` command that I'm using. You mentioned the possibility of false positives. The likelihood of an `.html` file in the codebase containing the string `api-uri-template` and the string `api/v` are pretty rare. That said, the next version of this code will very likely want to use a real HTML parser, which would remove a lot of the risk of false positives.

jmaldonado-okta commented on scripts/lint.sh:
```diff
@@ -0,0 +1,94 @@
+    interject 'Done installing dependencies'
+    interject 'Building Jekyll website'
+    # Build the Jekyll website
+    bundle exec jekyll build
```
nit: This is the actual build. Thus, consider moving this part to a different function or perhaps a more descriptive name than `setup` ?

joelfranusic-okta commented:
@jmaldonado-okta and @rlucero-okta thanks for your detailed reviews. I've made changes based on your feedback and added comments where I didn't make any changes. Can you take another look at the code?

rchild-okta commented:
@joelfranusic-okta Which commits should I be looking at? 

FYI, usually we don't rebase until we're ready to merge (and it would be a rebase, vs. merging in all the latest commits which is what is happening here?). If you did want to pull in the latest changes, better to rebase off master and then push.

joelfranusic-okta commented:
@jmaldonado-okta @rlucero-okta @rchild-okta - I should have addressed all of your comments. Additionally, I've squashed this topic branch down to one commit, how does it look to you now?

rchild-okta commented on scripts/common.sh:
```diff
@@ -0,0 +1,73 @@
+    if ! command -v protractor > /dev/null 2>&1; then
+        interject 'Installing Protractor'
+        # http://www.protractortest.org/#/tutorial
+        npm install -g protractor
```
We should install this locally as well, i.e. add it to package.json like this:
https://github.com/okta/samples-python-flask/blob/master/package.json#L33

You can then expose it via an npm run script like this:
https://github.com/okta/samples-python-flask/blob/master/package.json#L15

Which means that you would be able to run the command with `npm run test:e2e` (although in this case better to just name it "test")

In general, all npm dependencies should be listed in package.json, and we shouldn't be using some special mechanism to install it globally on ci/the user's system.

rchild-okta commented on scripts/common.sh:
```diff
@@ -0,0 +1,73 @@
+        interject 'Done installing Protractor'
+    fi
+    interject "Running 'webdriver-manager update'"
+    webdriver-manager update
```
And in this case, webdriver-manager update would be handled by that test command as well, so you wouldn't need it here.

rchild-okta commented on scripts/common.sh:
```diff
@@ -0,0 +1,73 @@
+        return 0
+    fi
+}
+
```
nit: don't need this extra line at the end

rchild-okta commented on scripts/common.sh:
```diff
@@ -0,0 +1,73 @@
+    fi
+    interject "Running 'webdriver-manager update'"
+    webdriver-manager update
+    interject 'Done checking Protractor dependencies'
```
Also, was expecting to see a new package.json file (with jasmine-reporter from the last time I looked at this PR) + a command somewhere here to run `npm install`

rchild-okta commented:
lgtm - squash and 🚢
